### PR TITLE
Adding autocasting

### PR DESCRIPTION
**Pull request by beta-ziliani**
Perfection is the enemy of good (but comments are welcomed).

### CODE REVIEW

**Review by straight-shoota — COMMENTED** [review]
I think there are a couple of important aspects missing:

* Autocasting happens at method calls (this one is definitely necessary to understand the topic, it's not stearing towards unnecessary perfection 😄 )
* Number literals are autocasted if the *concrete value* fits the target type. Non-literal values can't be checked at compile time, so they can only autocaste when their *type* fits the target type.
* Add an example for ambiguity that makes autocasting fail (could take the one from the release post)

File: docs/syntax_and_semantics/autocasting.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Type autocasting
+
+At function calls, Crystal transparently casts elements of certain types when there is no ambiguity.
```
Comment (straight-shoota) [inline docs/syntax_and_semantics/autocasting.md:3]: Oh, I was wrong there. Autocasting is not limited to function calls. It also applies to assignments. Maybe I missed something else?
```suggestion
Crystal transparently casts elements of certain types when there is no ambiguity.
```

File: docs/syntax_and_semantics/autocasting.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+## Number autocasting
+
+When passing a number to a function, values of a type are autocasted to a larger one if no precision is lost:
```
Comment (straight-shoota) [inline docs/syntax_and_semantics/autocasting.md:7]:
```suggestion
Values of a numeric type autocast to a larger one if no precision is lost:
```

File: docs/syntax_and_semantics/autocasting.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  x
+end
+
+foo 0xFFFF_u16 # OK, an UInt16 always fit an Int32
+foo 0xFFFF_u64 # OK, this particular Int64 fit in an Int32
+bar(foo 1)     # Fails, casting an Int32 to a Float32 might lose precision
+bar64(bar 1)   # OK, a Float32 can be autocasted to a Float64
+```
+
+Number literals are always casted when the actual value of the literal fits the target type, despite of its type. Expressions are casted too (like in the last example above), unless the flag `no_number_autocast` is passed to the compiler (see [Compiler features](compile_time_flags.md#compiler-features)).
```
Comment (straight-shoota) [inline docs/syntax_and_semantics/autocasting.md:28]: I think the last sentence should be moved to a separate paragraph, it's unrelated to autocasting of literals.

File: docs/syntax_and_semantics/autocasting.md (hunk 1 above)
Comment (beta-ziliani) [reply docs/syntax_and_semantics/autocasting.md]: I added an example for the other situation where this works.

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**